### PR TITLE
feat(ios, messaging): Allow notifications in foreground on iOS, configure in firebase.json

### PR DESCRIPTION
**Pull request by ken0nek**
### Description

Resolve this TODO comment (`"in a later version allow customising completion options in JS code"`).

Added a key to configure `UNNotificationPresentationOptions`
* [UNNotificationPresentationOptions | Apple Developer Documentation](https://developer.apple.com/documentation/usernotifications/unnotificationpresentationoptions)

### Related issues

* [🔥[🐛] Messaging: willPresentNotification delegate has no effect on iOS · Issue #5884 · invertase/react-native-firebase](https://github.com/invertase/react-native-firebase/issues/5884)
* [feat(ios): Add `banner` and `list` options to `IOSForegroundPresentationOptions` by ken0nek · Pull Request #463 · invertase/notifee](https://github.com/invertase/notifee/pull/463)
* [feat(ios): Add iOS 14 compatibility for `alert` by ken0nek · Pull Request #465 · invertase/notifee](https://github.com/invertase/notifee/pull/465)
* [feat: deprecate foregroundPresentationOptions.alert by helenaford · Pull Request #467 · invertase/notifee](https://github.com/invertase/notifee/pull/467)

### Release Summary

Can configure `UNNotificationPresentationOptions` on iOS via `firebase.json`

### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [ ] `Android`
  - [x] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [ ] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No

### Test Plan

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

### CODE REVIEW

File: packages/messaging/ios/RNFBMessaging/RNFBMessaging+UNUserNotificationCenter.m [hunk 1]
```diff
@@ -82,6 +83,50 @@ - (void)userNotificationCenter:(UNUserNotificationCenter *)center
...
+        presentationOptions |= UNNotificationPresentationOptionList;
+      } else {
+        // for iOS 13 we need to set `alert`
+        presentationOptions |= UNNotificationPresentationOptionAlert;
+      }
+    }
+  } else if (alert) {
+    // TODO: Remove `alert` once iOS 14 becomes the minimum deployment target
+    presentationOptions |= UNNotificationPresentationOptionAlert;
+  }
```
Comment (ken0nek) [inline packages/messaging/ios/RNFBMessaging/RNFBMessaging+UNUserNotificationCenter.m:128]: Same logic in `notifee`

Comment (mikehardy) [reply packages/messaging/ios/RNFBMessaging/RNFBMessaging+UNUserNotificationCenter.m:128]: yes - I was silently reading those as they went through, so I followed things as they went to the final solution, this looks good to me

**Review by mikehardy — APPROVED** [review]
Looks good to me, thanks a lot for this! Normally I tell people to use Notifee for device-local-anything when it comes to notification presentation but including at least a minimum for foreground (similar to background) is very nice.

One quick question: is there a feature gap on the android side as well for notifications showing up in the default state?

**Comment by ken0nek** [comment]
> One quick question: is there a feature gap on the android side as well for notifications showing up in the default state?

I suppose no.

As you can see [this doc](https://notifee.app/react-native/docs/ios/remote-notification-support), foreground presentation options for Remote Notification is only for iOS.

**Comment by ken0nek** [comment]
Hmm, I got no diff even after executing `yarn lint:markdown` 🤔

**Comment by mikehardy** [comment]
> I suppose no.
> 
> As you can see [this doc](https://notifee.app/react-native/docs/ios/remote-notification-support), foreground presentation options for Remote Notification is only for iOS.

Hmm - I think of Notifee as going "higher" in the stack, we are below them (in my mental model - that is, lower level) and firebase-android-sdk would be even lower than that, with the GCM libraries and FCM receipt on the device as the bottom / most inner part.

My question was aimed at: as GCM/FCM hits the device and firebase-android-sdk executes things / potentially booting the app and loading the Java code then the javascript bundle (similar to firebase-ios-sdk loading the objective-c code in `@react-native-firebase/messaging` that you changed) - if the app is in the foreground is a notification actually posted ever? Should it be possible in the way you added this feature here?

I have not actually tested this (at least not recently enough to remember) so I'm unsure - I was mostly curious for the results of any testing if you've done it. I'm okay with just doing an iOS feature and waiting for someone interested (or myself) to do same for android if there's a feature gap, I was just curious if you knew since I assume you're testing this with a react-native app so you looked at iOS and could probably check android with same codebase / testing setup

**Comment by mikehardy** [comment]
> Hmm, I got no diff even after executing `yarn lint:markdown` thinking

I see this:

---
```
mike@bistromath:~/work/Invertase/react-native-firebase (work/UNNotificationPresentationOption-config) % yarn lint:markdown -w
yarn run v1.22.18
$ prettier --check "docs/**/*.md" -w
Checking formatting...
[warn] docs/messaging/usage/index.md
[warn] Code style issues fixed in the above file.
Done in 3.44s.
mike@bistromath:~/work/Invertase/react-native-firebase (work/UNNotificationPresentationOption-config *) % git diff
diff --git a/docs/messaging/usage/index.md b/docs/messaging/usage/index.md
index a93f73e67..1383dd45a 100644
--- a/docs/messaging/usage/index.md
+++ b/docs/messaging/usage/index.md
@@ -395,13 +395,7 @@ Refer to [UNNotificationPresentationOptions](https://developer.apple.com/documen
 // <projectRoot>/firebase.json
 {
   "react-native": {
-    "messaging_ios_foreground_presentation_options": [
-      "badge",
-      "sound",
-      "alert",
-      "list",
-      "banner"
-    ]
+    "messaging_ios_foreground_presentation_options": ["badge", "sound", "alert", "list", "banner"]
   }
 }
 ```
```

**Comment by mikehardy** [comment]
(I pushed a commit with that diff to your branch as an attempt at higher velocity - hopefully no offense is taken! :-) )

**Comment by ken0nek** [comment]
Thank you for updating the doc!

> I was mostly curious for the results of any testing if you've done it.
> I was just curious if you knew since I assume you're testing this with a react-native app so you looked at iOS

I tested this feature by patching `@react-native-firebase/messaging` and `@react-native-firebase/app` in my company codebase. It works on iOS by sending a remote push notification, plus I got a notification on Android too.

> I'm okay with just doing an iOS feature and waiting for someone interested (or myself) to do same for android if there's a feature gap

Maybe we can explore the code around `onMessageReceived ` on Android in the future?

* [react-native-firebase/ReactNativeFirebaseMessagingService.java at main · invertase/react-native-firebase](https://github.com/invertase/react-native-firebase/blob/main/packages/messaging/android/src/main/java/io/invertase/firebase/messaging/ReactNativeFirebaseMessagingService.java#L35)
* [Receive messages in an Android app  |  Firebase Documentation](https://firebase.google.com/docs/cloud-messaging/android/receive)

File: .spellcheck.dict.txt [hunk 2]
```diff
@@ -2,15 +2,15 @@
 64k
 Ad
 AdMob
-Analytics
+AirPods
```
Comment (ken0nek) [inline .spellcheck.dict.txt:5]: Sorted alphabetically

File: .spellcheck.dict.txt [hunk 3]
```diff
@@ -168,30 +175,24 @@ timezones
 triaging
 TypeDoc
 UI
-UIViewController
 uid
+UIViewController
 uncomment
 unhandled
+UNNotificationPresentationOptions
```
Comment (ken0nek) [inline .spellcheck.dict.txt:182]: Added

**Review by mikehardy — DISMISSED** [review]
looks even better now, thanks for organizing + patching up the spellcheck dictionary

**Comment by mikehardy** [comment]
:thinking: perhaps this collided with #6406 and this needs a rebase against current main ?

**Comment by ken0nek** [comment]
> 🤔 perhaps this collided with #6406 and this needs a rebase against current main ?

Rebased with the latest main 👍

**Comment by mikehardy** [comment]
I'll be shocked if this doesn't do it - I'm about to be done for the night so won't see the result for hours but I imagine it will be either all green or perhaps a flaky-CI-test or something. But it all looks good now to me

**Pull request merged**